src/OpenTelemetry.AutoInstrumentation/Loading/AssemblyLoadDetector.cs

```diff
@@ -0,0 +1,68 @@
+// <copyright file="AssemblyLoadDetector.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+
+namespace OpenTelemetry.AutoInstrumentation.Loading
+{
+    internal abstract class AssemblyLoadDetector
+    {
+        private HashSet<string> _requiredAssemblies;
+        private int _loadedCount;
+
+        public AssemblyLoadDetector(IEnumerable<string> requiredAssemblies)
+        {
+            _requiredAssemblies = new HashSet<string>(requiredAssemblies);
+
+            AppDomain.CurrentDomain.AssemblyLoad += CurrentDomain_AssemblyLoad;
```

Comment:
Adding an `AssemblyLoad` handler per each detector would be inefficient. We should have one handler which works against some kind of `Dictionary<string(assemblyName), AssemblyLoadDetector>`.

I think you can change the `LazyInstrumentationLoader` so that `_assemblyLoadDetectors` will be a dictionary initialized when a detector is subscribed. After everything is ready you can add a single `AssemblyLoad` handler. The handler should be also removed when `LazyInstrumentationLoader` is disposed.

Reply:
I'm also concerned about having multiple handlers being registered and that we are not also unregistering the handlers. 

Perhaps this class needs to be its own thing and not part of an inheritance hierarchy. And this class just focuses on listening to assembly load events, and maintaining a collection of lazily loadable instrumentation loaders. This is similar to what @pellared mentioned above except that it assumes that the `LazyInstrumentationLoader` will not also be responsible for handling the assembly load events.

Reply:
https://github.com/open-telemetry/opentelemetry-dotnet-instrumentation/blob/10f9bbcf795c81670fe619ac5a78b21b9fedc86f/src/OpenTelemetry.AutoInstrumentation/Loading/AssemblyLoadDetector.cs#L52

Deregistration is actually here.

I'm more concerned if it becomes synchronized. The instrumentation instances need to be created asap after assembly load. Synchronized loop could cause some delays. That could be changed to parallel looping, but I'm not sure if that is actually cheaper.

Reply:
> I'm more concerned if it becomes synchronized

If I'm not mistaken, the event callback for `AppDomain.CurrentDomain.AssemblyLoad` is already synchronized, so if we loop over a list in one handler or invoke all the handlers there shouldn't be a difference. My preference though would be to only have one AssemblyLoad handler that handles all of the "detectors", like what @nrcventura is saying. This could make the design simpler and it could make possible some optimizations (for example, skipping processing of all the handlers if it's a well-known assembly that won't provide integrations)

Reply:
Seems you are right. I don't have currently newer reference but expect it's not changed in [AppDomain.cs](https://referencesource.microsoft.com/#mscorlib/system/appdomain.cs,3114).

So my idea would be then also to use one handler only for `AssemblyLoad` event, but the detector then should be changed to async event handling, to make sure that creating the instrumentation instances are not causing delays to each other. What do you think? @zacharycmontoya